### ISSUE TITLE
Windranger Focus Fire not resuming attack on target after losing vision

### ISSUE THREAD

**Issue by zetsupetsu** [jeffhill/Dota2#21]
Windranger's Focus Fire behaviour used to have her resume attacking the target once she regains vision of the focus fired target.
Currently, she stops attacking the target once she loses sight even for just a short time (juking in trees, invi, etc)

She used to be a good counter for PL/Manta as she always attacks the real one once Focus FIre has been locked, but in the recent patches this behaviour isnt there anymore and she just stands there or walks not attacking even if Focus FIre is active.

**Comment by jeffhill** [comment]
This will be fixed in the next hotfix. Thank you for the bug report!

**Comment by Aluerie** [comment]
@jeffhill thanks for the fix

I'm sorry for hijacking but just want to ask while track is "hot". 

Can you look at #1102, at both comments - Indicator and arrows model - It got broken after you started work on this issue so I suspect it *might be* related to this fix. It is still broken on just updated client.

**Comment by jeffhill** [comment]
Sorry, super unlikely to be related to this fix. I grabbed the ticket, then got busy with bugs from the update, and only actually did typing on this today.

**Comment by Aluerie** [comment]
@jeffhill I'm sorry again, but it seems like the something is still wrong in WR versus PL and even vision mini-game https://youtu.be/tjAgYURShQg

1) PL uses W - Wr attacks the closest Illusion even tho real "marked with focus fire" PL is still in attack range
2) with standard autoattack on - she starts turning on attacking closest dummy targets which is debatable but maybe it is weird
3) same example but with vision

**Comment by jeffhill** [comment]
The patch with the fix only went live at 4:46pm local time (1:15 ago), so after your post. Can you confirm it's not working with the current build?

**Comment by Aluerie** [comment]
@jeffhill I can confirm, my comment still stands. 

Showcasing 1) again (but a bit better) https://youtu.be/dS6eg6KdNTE - in my initial video it could seem like somewhat edge of attack range. Points 2) and 3) still stand. 

However, I should admit - I am not sure if my points should be considered as a "bug".

PL uses W - then illusions come out "first" in some "server queue of actions" -> WR with standard auto-attack starts attacking them as viable target > does not switch to Real PL because have automated standard attack order still on -> Real PL gets ignored. Same logic can be said about 2) and 3). 

So WR needs to move around to get around this Standard auto-attack logic getting in the way. I do not know, it just looks weird, It works fine with manta. 

It is kinda same problem of getting around fundamental game mechanics with other things like level 25 invisibility on Windrun talent: there is an annoying aspect that she stops Focus Firing because invis cancels autoattack order in its nature. So as I said I might be overthinking and this is not a bug - just clunky interaction with auto-attack behavior.

By the way, I thought the fix is live because fog vision part of issue seemed to work correctly

**Comment by jeffhill** [comment]
Thank you so much for your diligence here! Yes, this was my misunderstanding. I **only** changed the behavior when WR is idle, not when she's acquired a different attack target, while still being "active" on the focus fire target. I'll fix this now,

**Comment by jeffhill** [comment]
Ok, I've got another iteration on this - it'll ship in a hotfix in a few hours (this afternoon). I'm going to leave this issue open for now, in case more changes are needed.

**Comment by LokeLok** [comment]
Are you aware of this one? #1102 the Arcana icon is missing after recent update. Might want to include that one, while you at Focus Fire.

**Comment by Turmfalke2** [comment]
scroll up a bit he said this is super unlikely to be related and I agree with that assessment. One is broken cosmetic and one is actual game behavior. The aim here is to have one ticket per issue, not one ticket per hero or ability.

**Comment by LokeLok** [comment]
Im not sure how i missed the comment above, that's on me!

**Comment by Aluerie** [comment]
JeffHill, another suspicious/debatable interaction https://youtu.be/MIwhyLSdvPU

WR uses Focus Fire on AM -> WR herself leaves AM in fog of war -> orders some Move commands while not seeing the target -> comes back to high-ground, sees AM but does not continue focus-firing unless extra Attack order was given

**Comment by jeffhill** [comment]
Thank you, that repro case is super helpful! The transition from "move order" to "attack focusfire target" is connected now (it'll ship in a hotfix this afternoon), so the case you describe of having to wait for the move order to complete should be fixed.

Once again, I'll keep the ticket open for a bit in case anything else is discovered after the hotfix.

**Comment by jeffhill** [comment]
This hotfix is now live.

**Comment by VodkasRidge** [comment]
Dear Jeff!

First off all, thank you so much for fixing this issue! All the previous ult interactions with losing vision seem to be working fine now, however I have noticed a few new issues with movement when losing vision of a focus fire target. (And focus firing at a target in general) Again, thank you so much for all the hard work, and I hope this doesn't discourage you, as it seems hard to stamp out all these pesky bugs! I will list the issues in order of my perceived importance as a longtime WR spammer.

**Issue #1**

Description: When losing and regaining vision of a focus fire target anywhere on the map, Windranger will immediatly move to attack them. This will cancel any ongoing move commands, running to attack the focus fire target instead. 

To reproduce: Equip enemy hero with manta style > focus fire enemy hero > issue move command past the enemy hero > enemy uses manta style while WR is moving > WR will immediatly stop moving, will instead attack-move onto the enemy

This issue might seem small, however the same thing happens when losing vision in other ways. The issue seems to only occur if WR loses vision of the enemy and **then regains vision**.

To reproduce: Position enemy hero close to tree > focus fire enemy hero > issue a move command that makes WR move past the tree so she loses vision of the enemy and then regains it > as soon as WR regains vision of the enemy the move command is canceled and WR moves to attack the enemy instead

This issue seems to occur regardless of the enemy being in attack range or not, and regardless if WR is moving or standing still, for instance a focus fire target moving in and out of range of a ward on the other side of the map will make WR run to attack them immediately.

The main problem with this issue is that it makes running away while focus firing an enemy through the trees or similar almost completely impossible as WR will turn around everytime she loses and regains vision of the enemy resulting in you spinning on the spot. 

**Issue #2**

Description: Casting a spell or using any item will cause WR to immediatly run to attack her Focus Fire target afterwards

To reproduce: Focus fire enemy > blink away from the enemy (while keeping vision of the enemy) > WR will run to attack the enemy immeadiately after blink.

The same occurs when casting a spell such as powershot. The issue occurs regardless of distance on the map, for instance TP-ing to another lane and casting powershot there will still cause WR to run across the map to attack the focus fire target, provided she has vision of them. The issue occurs even if auto attack is set to "never".

I think these 2 issues mainly stem from the fact that moving and regaining vision somehow issues an attack-move command on WR's focus fire target. In fact however, Focus fire should never result in attack-move commands being issued, since WR should'nt **move to attack the enemy** she only should **attack the enemy while it's in range**. Sry for the long post.

**Issue #3** 

Description: While Focus Fire is active WR will not execute any shift-queue commands.

To reproduce: focus fire enemy target > issue shift-queue move commands, blink, etc. > WR will simply stand still, not executing any queued commands.

**Issue #4 (low-prio, cosmetic only)**

Description: While focus fire is active, WR will constantly play her running animation, even when not in range of the enemy hero.

To reproduce: Focus fire enemy target > move out of attack range > WR will run on the spot

I will update if I find any more of these

**Comment by VodkasRidge** [comment]
Actually after further testing **Issue 1** will still occur if vision of the Focus Fire target is lost without being regained as long as any enemy is in range.

**To reproduce:** Position enemy A and B next to each other > purchase Shadow Blade on enemy A > Focus Fire enemy A > issue a move command past enemy A and B > use Shadow Blade on enemy A > WR will stop moving immediately and attack enemy B

**Comment by VodkasRidge** [comment]
Somehow the patch just now seems to have fixed every single one of these issues! Thanks you whoever did this!

**Comment by VodkasRidge** [comment]
I have found some new (small) bugs that shipped with the last patch. I will list them in order of (my perceived) priority.

**(New) Issue 1**

**Description:** Using certain items or ground target attack while Focus Firing at a target will put Windranger in an incorrect attacking state that does not allow her to move and attack at the same time.

**To reproduce:** Purchase Blink Dagger on Windranger > Focus Fire an enemy target > blink within Focus Fire attack range of the enemy target > **WR can now only attack while standing still until a new attack command is issued directly on the enemy (ground target attack command will not work).**

The same will occur when using Shadow blade and using a ground target attack command to resume attacking your Focus Fire target.

**To reproduce:** Purchase Shadow Blade on Windranger >  Focus Fire an enemy target > Use Shadow Blade > issue a ground target attack command while in attack range of the Focus Fire target > WR can now only attack while standing still until a new attack command is issued directly on the enemy (ground target attack command will not work).

The same issue occurs when resuming attack after stopping focus fire either via casting a spell (while auto attack is set to "standard") or issuing a ground target attack command.

**To reproduce:** Focus Fire an enemy target > issue a stop command > issue a ground target attack command while in attack range of the Focus Fire target or cast Powershot (this will only work if Autoattack is set to "standard") > WR can now only attack while standing still until a new attack command is issued directly on the enemy (ground target attack command will not work).

I think these new issues are caused by attack commands especially when issued on the ground not working the same as right click attack commands issued directly on the enemy. Similar to the ground target attack command there seems to be the same issue with attack commands issued by the autoattack function that triggers after certain item effects and spell casts (if set to "standard"). It seems that Focus Fire only works correctly by the type of attack command that is issued by right clicking an enemy but not for the other types.

Note: if auto attack is set to "always" simply issuing a stop command while in range of the Focus Fire target will put WR in the described state of only being able to attack while standing still.

**(New) (minor) Issue 2**

**Description:** Using a TP scroll or Blink Dagger to move out of range of a Focus Fire target will cause WR to not attack the Focus Fire target when moving back into range.

**To reproduce:** Purchase Blink Dagger on Windranger > Focus Fire an enemy target > blink outside of Focus Fire attack range of the enemy target > move back into attack range > WR will not attack the target unless a new attack command is issued directly on the enemy (issuing a ground target attack command will put WR in the state described in Issue 1)

**Expected interaction:** I'm not 100% sure if this is intended, but it seems WR should start Focus Firing the target immediately upon re-entering attack range, only a stop command or going invisible should stop WR from attacking.

**(New) (minor) Issue 3**

**Description:** After completing a move command inside attack range of a Focus Fire target WR will always turn towards the target.

**To reproduce:** Spawn enemy hero Medusa > Focus Fire Medusa > Medusa casts Stone Gaze > issue a move command for WR away from Medusa while staying in attack range> WR will not get petrified while moving, however after completing her movement WR will turn towards Medusa and get petrified.

For clarification I am not talking about WR's **model**, everything works fine with that although there is some kind of glitchy/choppy animation that occurs while WR is "turning" after completing her movement.

**Expected interaction:** WR keeps her orientation in the direction she moved after completing the move command. Again, not 100% sure if this is intended, but before it was possible to stand still facing away from Medusa while still attacking her during Stone Gaze. Also the turning messes with items such as Force Staff, causing you to Force yourself at an angle or even toward the enemy if a move command away from the enemy is completed before you use the item.

Thanks for the hard work fixing these bugs!

**Comment by jeffhill** [comment]
Once again, thank you for all the repro cases and detailed steps! I should share a little of the backstory to the move while shooting stuff to explain why it's so temperamental.

The interaction of focus fire (and static link) with the state machine of the unit (the tiny AI that executes your orders) is very subtle, because units were never intended to do this sort of "shoot while moving" behavior. There's a "I am a unit and I am shooting" state, and a "I am a unit and I am moving over there" state and a "I am a unit and I am moving towards a target to shoot it" state, but these are all different things. We sort of repurposed the "moving towards a target to shoot it" state to also be the "I am moving *and* shooting at the same time" state as well.. but of course the relationships between states like "I am a unit and I am idle... oh look, it's my focus fire target, now I should shoot it but I should not *chase* it, I should continue to stand here"... these interactions are where the challenges are. Having the repro cases laid out precisely as you've been doing allows me to get into *exactly* the same state, repro the behavior, and discover what state transition is wrong in that case.

The good news is that there's only about 20 or so states total, so as we sort out the expected behavior in each of these interactions, the pool of potentially broken transitions only gets smaller.

I've just checked in another iteration of the focus fire state stuff, which should address everything *except* the "running while focus firing" behavior. That's an interaction of the state machine and the animation system I'd like to sort out, but only once all the gameplay behavior is finally "in the tin". I feel like having WR "run in place" is a small visual bug (she does look goofy though), while having her not shoot her focus fire target when they come into vision is a bug that might make a player quite grouchy.

Continuing to keep this open - because I bet there's going to be another weird interaction to discover. It wouldn't be Dota if there weren't just one more weird interaction. ;)

**Comment by jeffhill** [comment]
This hotfix is now live.

**Comment by opomuc** [comment]
> It wouldn't be Dota if there weren't just one more weird interaction.

and here goes Rubick with Focus Fire... :)

**Comment by Aluerie** [comment]
hey @jeffhill ; some issues mentioned by VodkasRidge are still there. 

1) [VodkasRidge's comment Issue 2](https://github.com/jeffhill/Dota2/issues/21#issuecomment-1158468464)  - my video: https://youtu.be/iWwHaw_VjuI

WR runs back to FF target after using spells/items with Standard AutoAttack option on. 

I use FF on PL > then blink, and cast spells/items on Naga and you can see that WR after each spell start running towards PL even in last showcase where I was attacking Naga for some time.

---
Following issues might be out of the scope of *this github issue* cause it is not directly about Focus Fire and target but definitely worth mentioning for you guys to get philosophy of mechanic "move+hit" right. 

2) [VodkasRidge's comment Issue 3](https://github.com/jeffhill/Dota2/issues/21#issuecomment-1158468464) - my video: https://youtu.be/It3wq59HjKo

Shift Queuing actions just does not work while WR is focus firing. Would be handy for some move commands/items.

3) [VodkasRidge's comment (New) (minor) Issue 3](https://github.com/jeffhill/Dota2/issues/21#issuecomment-1158539873) - my video: https://youtu.be/QG-QREZmwsM

Animation does not properly show where hero is turned to. WR animation is faced to dusa but she is not turned to stone because she is running back and thus probably in reality is facing away.

WR should be 1) considered to be turned to Dusa all the time or 2) there should be an animation of WR running back and firing backwards in a cool style where she doesn't even look at Dusa

**Comment by VodkasRidge** [comment]
Hey Jeff, first of all i wanted to say thank you for the detailed explanation of Windrangers Focus Fire movement, I absolutely love this hero and this game and Windranger in particular has brought me so much joy over the years that I am just really, really happy to be able to do a small thing to help fix her bugs :)

On this note, it seems like Windrangers behavior was essentially rolled back to how it was in my first post, while new issue 1 and 2 seem to be fixed, the original issues 1, 2 and 3 all seem to occur exactly in the same way as described in my first post. I'll just copypaste the relevant sections here:

_Issue 1

Description: When losing and regaining vision of a focus fire target anywhere on the map, Windranger will immediately move to attack them. This will cancel any ongoing move commands, running to attack the focus fire target instead.

To reproduce: Equip enemy hero with manta style > focus fire enemy hero > issue move command past the enemy hero > enemy uses manta style while WR is moving > WR will immediately stop moving, will instead attack-move onto the enemy

The same thing happens when losing vision in other ways. The issue seems to only occur if WR loses vision of the enemy and then regains vision.

To reproduce: Position enemy hero close to tree > focus fire enemy hero > issue a move command that makes WR move past the tree so she loses vision of the enemy and then regains it > as soon as WR regains vision of the enemy the move command is canceled and WR moves to attack the enemy instead

This issue seems to occur regardless of the enemy being in attack range or not, and regardless if WR is moving or standing still, for instance **a Focus Fire target moving in and out of range of a ward on the other side of the map will make WR run to attack them immediately**.

To reproduce: Place Observer Ward > position enemy hero close to the edge of ward vision > Focus Fire enemy hero > Teleport WR to another lane/move WR far away > Move enemy out of and back into Ward vision > WR will move to attack enemy across the map


Issue 2

Description: Casting a spell or using any item will cause WR to immediately run to attack her Focus Fire target afterwards

To reproduce: Focus Fire enemy > blink away from the enemy (while keeping vision of the enemy) > WR will run to attack the enemy immediately after blink.

The same occurs when casting a spell such as powershot. The issue occurs regardless of distance on the map, for instance TP-ing to another lane and casting powershot there will still cause WR to run across the map to attack the Focus Fire target, provided she has vision of them. The issue occurs even if auto attack is set to "never".

To reproduce: Focus Fire enemy > move out of attack range of the enemy (while keeping vision of the enemy) > use power shot > WR will run to attack the enemy immediately after completing the channel.

I think these 2 issues mainly stem from the fact that Focus Firing at a target somehow issues attack-move commands on WR's Focus Fire target. In fact however, Focus Fire should never result in attack-move commands being issued, since WR shouldn't move to attack the enemy she should only attack the enemy while it's in range.

Issue 3

Description: While Focus Fire is active WR will not execute any shift-queue commands.

To reproduce: Focus Fire enemy target > issue shift-queue move commands, blink, etc. > WR will simply stand still, not executing any queued commands.

Also, any Shift-queue including a Focus Fire command will only work until the Focus Fire command is issued, any further queued commands will not work and the queue will stop in the same way as when trying to shift-queue commands while Focus Fire is already active._

Aside from these already described bugs, I could not find any new ones. 
Thank you so much for your time and attention :)

**Comment by Makakass** [comment]
Some other static link/focus fire attaking bugs:
https://github.com/jeffhill/Dota2/issues/1324 https://github.com/jeffhill/Dota2/issues/1325

**Comment by jeffhill** [comment]
The hotfix today will have what I believe are fixes for all these behaviors, except changing the animation for WR (that's a bigger content conversation). The interactions with auto-attack are subtle because it matters if you've clicked to attack the target or if you've clicked to move past the target (is this an "attack order" or a "move order" - in both cases you're shooting while moving...)

Thanks again for all your help with this!

**Comment by jeffhill** [comment]
This hotfix is now live.

**Comment by Aluerie** [comment]
1. WR still forgets about Move commands if she gets "confused" - disarmed or shadow realm mid-walking point. https://youtu.be/Djpz7OfnH4w

FocusFire > WR clicks Move > Enemies disarm her > WR stops for disarm duration > continues FocusFiring from stop point.

Expectation: continue walking. Similar thing was fixed for PL W/Manta/etc but it still exists for Hallberd/Book of shadows/Shadow Realm/ maybe more

2. WR still does not properly handle Taunts https://youtu.be/Djpz7OfnH4w

FocusFire PL > Axe uses Q on WR > WR stays still and does not attack anybody at alld

Expectation: WR starts attacking Axe. NOTE: LC gets E procs WITHOUT WR actually attacking her. even more it looks so fast that i suspect it considers FF attack speed

3. The issue about going back to FF target is still here but in a very weird form. some logic is not perfect https://youtu.be/d0_AgiFrbeo

I only press blink here, nothing else and WR wants to go somewhere after some blink points (not everywhere i blink to causes this behaviour)

**Comment by jeffhill** [comment]
FF vs. disarm fix checked in, and FF vs. taunt fix checked in - these will ship in the next hotfix (likely a little later today).

The FF target blink thing doesn't actually seem to be a FF thing, from what I can tell. I can repro it without using focus fire, just attacking the target. It *might* be a bug, but given that it's not FF-related, I'm really hesitant to touch it as part of this (since that code is hit by every single unit, all the time, FF or not). If it's problematic, submit a different issue and we can look at it systemically - is that fair? I'd like to actually be able to close this issue someday. ;)

As always keeping the ticket open for now, because Dota.

**Comment by bclaw524** [comment]
I had a match as WR where I used ult on an enemy Tiny that managed to disengage from me, however, my movement command became wonky and slow, like my hero was trying to stutter step the whole time. 

I lost and gained vision a few times but Tiny was already outside of my attack range. Whether I was trying to chase or move away from the target, my movement became slowed and WR kept "turning" to face Tiny.

Match ID: 6627449479, around 12:45~13:10 in the mid lane near Tier 1 Radiant

**Comment by jeffhill** [comment]
Thanks for the replay! It looks like you Focus Fired the Tiny but weren't in range - then got a stray shot in on the tower, and somehow your hero kept turning back to the tower when you lost vision of Tiny, despite clicking move orders off to the West. What do you have your autoattack setting set to?

I'll take a look and see if I can get a repro. Thanks!

**Comment by Aluerie** [comment]
hey, JeffHill ! I'm really sorry to terrorize you with this issue... Maybe, this issue should indeed be closed and moved on :c Even though this replay from bclaw524 is very interesting 🤔 

---

Anyway, other bugs/inconsistencies to this issue:
1. Fear interaction is different when there is Move order present and might lose some DPS for WR: https://youtu.be/i6xIPWgfNuE

Repro: FocusFire PA -> WR orders Move somewhere -> Gets Feared -> She does not continue attacking PA after Fear

Expectation: She should continue attacking PA. It works fine pretty much in all other scenarios, normal/weird/exotic (we discussed many above), including if WR stands still while FFing and then gets Feared.

2. Revealing Invisible Target with placing sentries/picking up gem confuses WR: https://youtu.be/pRlsgURyG8E

Repro: FocusFire PA -> PA presses SilverEdge -> WR uses Sentry/picks up Gem -> WR does not continue attacking PA

Expectation: continue attacking PA. This might be debatable but again, it works fine for other cases, such as ally using sentry/gem (WR player can even give move orders just fine while target is invis). WR using Dust herself also works fine. Target going in and out of sentry range. Etc.

3. I found 3 spells that confuse WR as well, this might be not even related to FF but anyway: https://youtu.be/ltaVW1nNpYo

Repro: WR FocusFires DW -> WR orders Move command somewhere ->
and now either of a) b) c):
a) DW uses W while being FF'ed
b) Ally Pudge eats WR with the Shard
c) Ally/Enemy Kunkka uses X Mark The Spot on WR
-> WR gets confused and does not continue attacking the FF target automatically (like in all other cases we discussed above and that were fixed)

Expectation: WR continues attacking

---

1 and 3 (maybe 2 as well) have a common factor of WR having Move order on her mind. So maybe, again, some logic is not perfect, but either way thanks a lot, Jeff. And I really really feel bad and sorry for all of it :x

**Comment by jeffhill** [comment]
The "follow while FF" issue was tricky - I ended up solving it by making "follow friendly unit" orders turn into "move to position" orders during focus fire attacking. If you're actively shooting at your focus fire target and you right click on a friendly, it's just going to assume you meant "move here" rather than "follow this NPC", because there's no state for the unit to follow an NPC *and* be focus firing something else. It didn't seem worthwhile to take on the added risk and complexity of adding a new unit state for that, since it didn't sound particularly desirable? It feels like the most common case of clicking on an ally during FF is when you accidentally click on them while moving around, so this will just do "what you meant to do."

I didn't touch the fear thing, specifically, though the fix for 3 might fix it now that I'm typing this. It seems fine to me if fear breaks focus fire targeting and you need to click on your target again post-fear? It also seems fine if it doesn't work that way either... it's hard to feel too strongly about the behavior in that case since fear is so disruptive overall to your hero, there are so many reasonable outcomes for what to do once fear is over, and realistically the player's going to be issuing orders the whole time anyhow.

Focus fire vs. revealed target was just the idle state not respecting that the FF target has become available, so that's fixed.

I couldn't repro focus fire losing tracking vs. Dark Willow's Shadow Realm, so either it's fixed with these other changes or maybe I've missed the repro steps somehow? That seems maybe worthwhile to fix if it's still happening after the hotfix. The pudge vs. WR interaction seems even more niche than the fear interaction and it's hard to feel that it's really wrong or broken in that case?

These fixes will be in the hotfix later today. It's been a long road on this issue, but I'm feeling good about this hotfix - it's possible today's the day we get to close this issue. **fingers crossed** Thanks for the feedback on the changes, it's all really good stuff that's getting fixed here!

**Comment by Aluerie** [comment]
All your remarks are fair! Hotfix is live so... thanks for fixing 1 (this fix idea is perfect!) and 3. Both Pudge/Kunkka cases from 4 are fixed. And weirdly enough, Fear inconsistency is actually fixed right now. Now Fearing both moving and stationary WR with FF on behaves in a same way - great ! So in the end everything except DW was addressed.

--- 
DW case is still a bit inconsistent with everything from above but can be considered *fixed* ! Maybe it is niche enough to call it a "feature" and close the issue :D

Hopefully very last Repro: https://youtu.be/Ie2U0Wt7Qdo
FocusFire DW -> (!) WR orders Move somewhere 
-> either a) or b) while WR is still in Move+Attack state
a) DW uses W 
b) New discovery: Bane uses Sleep on DW
 ->WR stops at place -> Shadow Realm/Sleep is over -> WR continues attacking

Question: Should WR continue moving after DW W/Sleep usage ? WR continues to move for other cases of "confusion" like disarm/ghost/astral. Note: Bane interaction sometimes behaves correctly and sometimes does not, I can't figure out why.

**Comment by zeroclona** [comment]
I 

> @jeffhill I'm sorry again, but it seems like the something is still wrong in WR versus PL and even vision mini-game https://youtu.be/tjAgYURShQg
> 
>     1. PL uses W - WR attacks the closest Illusion even though real "marked with focus fire" PL is still in attack range/vision
> 
>     2. with standard auto-attack on - she starts turning on attacking closest dummy targets which is debatable if it is correct behavior (most likely it is) but maybe it is weird
> 
>     3. same example but with vision
> 
> 
> Even though technically WR just needs to move to avoid this bug/interaction but it still can cost some seconds of damage :D

Cmon, its intended to juke with doppel or manta or mirror image or etc enemies. WR should not find real PL instantly after casted doppel. 
Obviously its ok if player himself click the right one.
After these hotfixes its impossible to play on PL vs WR
A long time ago it worked like now. And this bug was fixed in PL's favor

**Comment by jeffhill** [comment]
So it has been the case for a while that WR with focus fire could A-click the ground and would attack the focus fired target. Since illusions don't duplicate or purge the focus fire debuff, this meant you could pretty easily start shooting the "real" PL again. We've just changed the gameplay code around illusion creation to be considered similar to a stop command, for the purposes of Focus Fire. We've also change the focus fire resume behavior to resume when WR starts an attack on the FF'ed target, rather than based on an attack or attack-move order being issued.

The net result of this is much more consistent behavior. If you focus fire a target, then hit S, then attack a different target using any of the attack methods, WR will consistently attack that other target (rather than snapping to the FF'ed target). Similarly, if you are attacking a FF target who then mantas, the completely normal attack target selection logic from every other hero applies to Windranger - and she only resumes the new fire on the move behavior when she **actually attacks** the FF target.

This change will ship in the next hotfix and will hopefully finally close this issue.

**Comment by jeffhill** [comment]
This is now live. I'm going to close this one now, but if there's more to be done here, please post and we can re-open it.

**Comment by Aluerie** [comment]
Now second comment - list of bugs.

1. Cannot deny runes/items while FF is on

    [YT Repro](https://youtu.be/oHqrtgTCT7s) WR uses FF on PL -> tries to attack Rune/dropped item -> WR gets stuck in place

    **Expectation.** WR should attack these objects

2. Idle state still finds real PL

   [YT Repro](https://youtu.be/1gL1u5unvlM) WR uses FF on PL -> WR does not move while PL uses W -> WR continues attacking real PL 

    **Expectation:** if we really keep this iteration of FF behavior then she shouldn't find the real PL.

3. niche/debatable WR can confuse herself with issuing attack and then move on allies
    
    [YT Repro](https://youtu.be/Y3m2C1BQqOg) WR uses FF on PL -> WR issues Attack Order on ally BS -> WR issues Move Order on ally BS -> WR stops focus firing
    
    **Expectation:** After Failing Attack on BS she should do move+attack in her Move command and thus continue FocusFiring.  
    
4. #1632 
    
PS, feel free to ask me to shut the... Maybe, I'm violating the whole concept of "bugtracker" with stalling progress in other *no less important* issues.

**Comment by VodkasRidge** [comment]
First off, i wanted to second Aluerie on Issue 0 and add to it, secondly i found a bug.

1. To add to Aluerie (as a fellow WR spammer): Windrangers Focus Fire ability does not really seem to be designed around the player being easily able to tell **whether it is on or not**. While this issue is lessened when equipping an arcana the base WR set doesn't even have an animation for Focus Fire attacking vs normal attacking. To my understanding this is the case because the entire ability is designed around being **constantly turned on** after casting it on a target **once**. There is no need for the player to be able to check whether they are currently Focus Firing on a target because there simply shouldn't be anything that stops FF **aside from the player conciously interrupting the ability.** After casting Focus Fire the player should not have to worry about whether they are attacking or not, instead being able to focus solely on movement, and maintaing proper spacing to the target aswell as casting WRs other spells. I realize that this might not be the thread for this discussion, but i still feel that it fits here **because this is what the original issue was about,** WR randomly stopping Focus Fire (and why it caused so much grief).

If this is not the intended design for this ability then it should be **A:** very easy for the player to tell whether they are currently focus firing or not and **B:** very clear exactly which interactions aside from pressing "s" stop Focus Fire. 


**Bug:**  While having the "Compass of the Rising Gale" Arcana equipped cosmetic arrows from Focus Fire will remain in the target even after they use manta style or a similar ability, allowing the player to distinguish the FF target from the illusions.

**To reproduce:** Equip Arcana on WR --> WR uses FF on PL and hits him a couple of times --> PL uses Doppelganger --> the real PL will have arrows stuck on his model.

Edit: I see that it was not an intended balance change and my first point was already adressed. Thanks again for the hard work!

**Comment by Aluerie** [comment]
I can't even find any words to express my infinite gratitude for attention you give to this issue and my comments, thanks a lot for the fix, it really means a lot :smile:  

---

Everything I said in both comments except two things was fixed/cleared up, thus, I want to ask about these untouched points again with a bit better reproduction steps. 

1. Ground AttackMove does not work outside of FF acquisition range

[YT Repro:](https://youtu.be/KhhhkL-nEhc) PL and Naga stay far away from each other -> WR uses FF on PL -> WR runs towards Naga with Move/ground AttackMove -> WR orders ground AttackMove around Naga -> WR does not attack Naga

**Expectation:** WR should attack Naga

It seems like currently ground Attack-Move outside of FF range behaves the same way as just Move order. I hope it is not a *necessary* sacrifice to get all other cases right in Dota unit behavior code. I hope fix to this problem won't create any problem at *just outside of FF acquisition range* and 1b (Powershot>S>A small QoL). To add extra, even pressing S for Stop order does not help with ground AttackMove. The only way to attack Naga is to actually hover over Naga and press A on her model.
  
2. [niche] MoveFollow: Adding extra steps can break intended move+attack behavior

[YT Repro A:](https://youtu.be/YIts_qLk9UM) WR uses FF on PL -> WR orders MoveFollow on ally Pudge -> WR blinks away -> WR orders Move-Follow on ally Pudge again -> WR does not keep FFing PL

[YT Repro B:](https://youtu.be/NSx1uopt9nA) WR uses FF on PL -> WR orders Attack on full-hp ally Pudge -> WR orders Move-Follow on that Pudge -> WR does not keep FFing PL

**Expectation:** WR still should keep FFing PL in both cases. 

Just casual MoveFollow on ally Pudge works fine as intended by your fix last week but as you can see adding extra steps can ruin such behavior. It's quite niche interaction bcs WR player probably spams move command anyway however it might signal to some logic issues.

**Comment by jeffhill** [comment]
I understand how these behaviors are happening and I don't think they indicate any deeper issues, though I agree they're minor bugs. For now, I'd like to just leave that code alone for a little bit and see if any problems come to light - but thank you for reporting this! The next time we're iteration on the unit behavior code, I'll see about cleaning these corner cases up.

**Comment by Lyralei27** [comment]
The bug is back with the aghs only
